network/src/network.rs

```diff
@@ -337,26 +343,90 @@ impl NetworkState {
             .collect::<Vec<_>>()
     }
 
+    pub(crate) fn can_dial(&self, peer_id: &PeerId, addr: &Multiaddr) -> bool {
+        if self.local_peer_id() == peer_id {
+            trace!(target: "network", "Do not dial self: {:?}, {}", peer_id, addr);
+            return false;
+        }
+        if self.listened_addresses.read().contains_key(&addr) {
+            trace!(target: "network", "Do not dial listened address(self): {:?}, {}", peer_id, addr);
+            return false;
+        }
+
+        if self
+            .failed_dials
+            .lock()
+            .get_refresh(peer_id)
+            .map(|last_dial| last_dial.elapsed() <= DIAL_FAILED_WAIT)
+            .unwrap_or(false)
+        {
+            trace!(target: "network", "Do not dial address fialed less than 5 minutes: {:?}, {}", peer_id, addr);
+            return false;
+        }
+
+        let peer_in_registry = self.with_peer_registry(|reg| {
+            reg.get_key_by_peer_id(peer_id).is_some() || reg.is_feeler(peer_id)
+        });
+        if peer_in_registry {
+            trace!(target: "network", "Do not dial peer in registry: {:?}, {}", peer_id, addr);
+            return false;
+        }
+
+        if let Some(dial_started) = self.dialing_addrs.read().get(peer_id) {
+            trace!(target: "network", "Do not repeat send dial command to network service: {:?}, {}", peer_id, addr);
+            if dial_started.elapsed() > DIAL_HANG_TIMEOUT {
```

Comment:
consider to send this message to sentry?

Reply:
This will be called frequently, we will noticed as soon as possiable